packages/imperative/src/events/src/ImperativeEventEmitter.ts

```diff
@@ -0,0 +1,288 @@
+import { Logger } from "../../logger/src/Logger";
+import { LoggerManager } from "../../logger/src/LoggerManager";
+import { IImperativeRegisteredAction, IImperativeEventEmitterOpts, IImperativeEventJson } from "./doc";
+import { ConfigUtils } from "../../config/src/ConfigUtils";
+
+export class ImperativeEventEmitter {
+    private static mInstance: ImperativeEventEmitter;
+    private static initialized = false;
+    private subscriptions: Map<string, [fs.FSWatcher, Function[]]>;
+    private eventTimes: Map<string, string>;
+    public appName: string;
+    public logger: Logger;
+
+    public static initialize(appName?: string, options?: IImperativeEventEmitterOpts) {
+        if (this.initialized) {
+            throw new ImperativeError({msg: "Only one instance of the Imperative Event Emitter is allowed"});
+        }
+        this.initialized = true;
+
+        if (ImperativeConfig.instance.loadedConfig == null || LoggerManager.instance.isLoggerInit === false) {
+            ConfigUtils.initImpUtils("zowe");
+        }
+
+        this.instance.appName = appName;
+        this.instance.logger = options?.logger ?? Logger.getImperativeLogger();
+    }
+
+    public static teardown() {
+        if (!this.initialized) {
+            return;
+        }
+
+        for (const sub of [...this.instance.subscriptions.keys()]) {
+            this.instance.unsubscribe(sub);
+        }
+
+        this.initialized = false;
+    }
+
+    public static get instance(): ImperativeEventEmitter {
+        if (this.mInstance == null) {
+            this.mInstance = new ImperativeEventEmitter();
+            this.mInstance.subscriptions = new Map();
+            this.mInstance.eventTimes = new Map();
+        }
+        return this.mInstance;
+    }
+
+    /**
+     * Check to see if the Imperative Event Emitter instance has been initialized
+     */
+    private ensureClassInitialized() {
+        if (!ImperativeEventEmitter.initialized) {
+            throw new ImperativeError({msg: "You must initialize the instance before using any of its methods."});
+        }
+    }
+
+    /**
+     * Check to see if the directory exists, otherwise, create it : )
+     * @param directoryPath Zowe or User dir where we will write the events
+     */
+    private ensureEventsDirExists(directoryPath: string) {
+        try {
+            if (!fs.existsSync(directoryPath)) {
+                fs.mkdirSync(directoryPath);
+            }
+        } catch (err) {
+            throw new ImperativeError({ msg: `Unable to create '.events' directory. Path: ${directoryPath}`, causeErrors: err });
+        }
+    }
+
+    /**
+     * Check to see if the file path exists, otherwise, create it : )
+     * @param filePath Zowe or User path where we will write the events
+     */
+    private ensureEventFileExists(filePath: string) {
+        try {
+            if (!fs.existsSync(filePath)) {
+                fs.closeSync(fs.openSync(filePath, 'w'));
```

Comment:
Following up on the CodeQL failure, this could be addressed by doing an `openSync` as the first statement in the `try` block and then creating the file in the `catch` block (as it will fail to open if the file does not exist)

Reply:
I do like the approach as it will help prevent errors thrown due to the file not existing.
However, this technique takes advantage of the `-w` system flag, which will create the file if it doesn't exist or truncate it if it dow ([more details](https://nodejs.org/api/fs.html#file-system-flags))

Reply:
In the current logic, I think that openSync will not truncate an existing file because the earlier 'if' test should prevent openSync from being called when the file already exists. Admittedly, I do not know if other parts of this class cares whether existing files are truncated before use.